keps/sig-node/3542-cri-image-pulling-with-progress-notification/README.md

```diff
@@ -0,0 +1,876 @@
+the minimum amount of data the runtime needs to pull (as opposed to total / complete image size)
+in order to start the container should be considered the total amount of download being in progress.
+For instance, if the complete container image consists of 10 layers, 1 GiB per layer, and runtime
+can start the container with only 4 layers pulled, then 4GiB should be considered the total download
+amount and the progress message to Kubelet at hypothetical 1GiB progres point would look like:
+```
+{
+  image_ref: "registryA.com/repo/imageB:tagC",
+  offset: 1073741824,
+  total: 4294967296,
+}
+```
+When downloading the minimum needed layers to start the container is done, the progress reports
+stream can be closed and pull operation can proceed silently in the runtime without notifications
+sent to Kubelet for the rest of the image being lazily pulled as needed.
+
+If ImagePullProgress is enabled for kubelet but runtime does not support it, the fallback should
+use regular silent ImagePull API call.
+
+### Risks and Mitigations
+
+The risks are minimal, introduction of new API is not affecting any other APIs.
+
+Misconfiguration of image pull request granularity for kubelet can result in big number
+of progress responses published by runtime into return stream. This can be mitigated by kubelet
+throttling the amount of events it publishes to Pod object (publishing less than received from
+runtime), as well as in runtime by metering amount of responses being sent within time interval.
+Intention of this KEP's CRI extension is to provide not-so-frequent updates for Kubelet (e.g. once
+a minute), while having a possibility to have more-frequent updates for cli tools (e.g. once a
+second).
+
+Documentation should be updated to reflect new FeatureGates, Kubelet config options, and performance
+overhead for reporting progress from image pulls.
+
+Runtime can be such that does not support image pull progress reporting. In this case fallback to
+regular image pulling call should happen on client side (kubelet, cli tool, other entities
+requesting image to be pulled from runtime).
+
+## Design Details
+
+Following new CRI streaming API call is proposed:
+
+    // PullImageWithProgress pulls an image with authentication config.
+    // It returns periodically amount of image pulled so far.
+    rpc PullImageWithProgress(PullImageWithProgressRequest) returns (stream PullImageWithProgressResponse) {}
+
+The PullImageWithProgress() API call will connect and send a PullImageWithProgressRequest message to
+runtime CRI image service server.
+
+The PullImageWithProgressRequest contains base information needed to do the image pull (image name, auth config
+and sandbox information), and it will also contain information how often the server should send progress reports.
+The CRI client can restrict the progress reporting to be time-based (e.g. once every n. seconds),
+or based on size (amount of bytes/KiB/MiB downloaded). The size-based should be the default one.
```

Comment:
Time based one is more useful for detecting whether there is progress or not.

Reply:
I suppose we shouldn't restrict the runtime to chose the default, I'll remove that sentence.